### PR TITLE
#19 add more input model report command

### PR DESCRIPTION
**Pull request by MaXeraph**
#19 

## Changes
- Add nodemon and `dev` run scripts
- Report command will now popup a modal that allows for fields input
- Discord does not let us know if the user cancels modal submission, so if no submission within 1 minute, assume abort.
- The values in the screenshot below are automatically filled when the modal first pops up. 

## Screenshots
image

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -4,14 +4,14 @@
 
 "@cspotcode/source-map-support@^0.8.0":
   version "0.8.1"
-  resolved "https://registry.yarnpkg.com/@cspotcode/source-map-support/-/source-map-support-0.8.1.tgz#00629c35a688e05a88b1cda684fb9d5e73f000a1"
+  resolved "https://registry.npmjs.org/@cspotcode/source-map-support/-/source-map-support-0.8.1.tgz"
```
Comment (umar-ahmed) [inline yarn.lock:7]: We should revert these changes back to the yarn mirror

File: src/commands/report.ts [hunk 2]
```diff
@@ -54,3 +81,87 @@ export async function execute(interaction: CommandInteraction) {
     });
   }
 }
+
+function generateModal(user: any, options: any) {
```
Comment (umar-ahmed) [inline src/commands/report.ts:85]: I think we can type this a bit better. Let me know if you want some help with that, but basically you can hover over the value that's being passed in and see what type it is

File: package.json [hunk 3]
```diff
@@ -4,6 +4,7 @@
   "version": "1.0.0",
   "description": "The official ChainPatrol Discord bot",
   "scripts": {
+    "dev": "nodemon --watch 'src/**/*.ts' --exec 'ts-node' src/index.ts",
```
Comment (umar-ahmed) [inline package.json:7]: Nice!

File: src/commands/report.ts [hunk 4]
```diff
@@ -19,32 +27,51 @@ export async function execute(interaction: CommandInteraction) {
...
-        assets: [
+    const modal = generateModal(user, options);
+
+    // Show the modal to the user
+    await interaction.showModal(modal);
+
+    // extract data from modal
+    interaction
+      .awaitModalSubmit({ time: 60000 })
+      .then(async (modalInteraction) => {
```
Comment (umar-ahmed) [inline src/commands/report.ts:38]: Can we `await` this promise instead of chaining with `.then()` and `.catch()`?

File: src/commands/report.ts [hunk 5]
```diff
@@ -54,3 +81,87 @@ export async function execute(interaction: CommandInteraction) {
...
+    .setPlaceholder(`Please paste any screenshots URL related to the scam`);
+  const screenshotsActionRow =
+    new ActionRowBuilder<ModalActionRowComponentBuilder>().addComponents(
+      screenshotsInput
+    );
+
+  const contactInput = new TextInputBuilder()
+    .setCustomId("contactInput")
+    .setLabel("Let us know how to best contact you")
+    .setRequired(true)
```
Comment (umar-ahmed) [inline src/commands/report.ts:148]: not required

File: src/commands/report.ts [hunk 6]
```diff
@@ -54,3 +81,87 @@ export async function execute(interaction: CommandInteraction) {
...
+    );
+  const descripionActionRow =
+    new ActionRowBuilder<ModalActionRowComponentBuilder>().addComponents(
+      descriptionInput
+    );
+
+  const screenshotsInput = new TextInputBuilder()
+    .setCustomId("screenshotInput")
+    .setLabel("Screenshot URLs (One URL per line)")
+    .setRequired(true)
```
Comment (umar-ahmed) [inline src/commands/report.ts:137]: not required

File: src/commands/report.ts [hunk 7]
```diff
@@ -54,3 +81,87 @@ export async function execute(interaction: CommandInteraction) {
...
+    .setValue(`Discord Report: ${url}`);
+  const titleActionRow =
+    new ActionRowBuilder<ModalActionRowComponentBuilder>().addComponents(
+      titleInput
+    );
+
+  const descriptionInput = new TextInputBuilder()
+    .setCustomId("descriptionInput")
+    .setLabel("Description")
+    .setRequired(true)
```
Comment (umar-ahmed) [inline src/commands/report.ts:123]: not required

**Review by umar-ahmed — CHANGES_REQUESTED** [review]
Overall looks good. I haven't QA'd it yet, but there's some code changes I think you should make

File: yarn.lock (hunk 1 above)
Comment (MaXeraph) [reply yarn.lock]: https://github.com/chainpatrol/discord-bot/pull/23/commits/9e893e57519b8b20c2de90484c8e00bc12fb269b

File: src/commands/report.ts (hunk 2 above)
Comment (MaXeraph) [reply src/commands/report.ts:85]: Yeah I might need help with this, I wasn't able to just infer the type, IDE was not happy

File: src/commands/report.ts (hunk 4 above)
Comment (MaXeraph) [reply src/commands/report.ts]: https://github.com/chainpatrol/discord-bot/pull/23/commits/0d41ddac98242f6dbc5da7c1fcacd6ea1c3168a9

File: src/commands/report.ts (hunk 2 above)
Comment (MaXeraph) [reply src/commands/report.ts]: Attempt 1: https://github.com/chainpatrol/discord-bot/pull/23/commits/6b9914c106d5cda1007acffe224728ddb84626e1

**Review by umar-ahmed — COMMENTED** [review]
Code LGTM. Will QA it later today and approve if it looks good.

**Comment by umar-ahmed** [comment]
## QA Discussion

- Remove screenshot
- Add four digit identifier to Discord username (ex. `umariomaker#1234`)

**Comment by MaXeraph** [comment]
> ## QA Discussion
> 
>     * Remove screenshot
> 
>     * Add four digit identifier to Discord username (ex. `umariomaker#1234`)

https://github.com/chainpatrol/discord-bot/pull/23/commits/3c07dacddee8bc5d1229114c402824ce6c2d1ec5

**Comment by umar-ahmed** [comment]
✅ QA LGTM

CleanShot 2023-05-01 at 10 16 13@2x
CleanShot 2023-05-01 at 10 16 08@2x

**Review by umar-ahmed — APPROVED** [review]

**Comment by NikitaVr** [comment]
@MaXeraph @umar-ahmed Looks good to me, One non-blocking Suggestion would be to auto-generate the title when the url is finished typing if the title is empty.

**Review by NikitaVr — APPROVED** [review]

**Comment by umar-ahmed** [comment]
> @MaXeraph @umar-ahmed Looks good to me, One non-blocking Suggestion would be to auto-generate the title when the url is finished typing if the title is empty.

@NikitaVr That already happens based on the URL you input to the slash command.

**Comment by NikitaVr** [comment]
Got it no action needed then
